File: apps/api/src/main/java/com/protrack/shared/error/Problem.java
package com.protrack.shared.error;

import java.time.OffsetDateTime;
import java.util.List;

/**
 * Standardized error response body (RFC 9457-style Problem Details) as defined in the approved
 * REST API Specification (§1.5). Returned by {@link GlobalExceptionHandler} for every error.
 *
 * @param fieldErrors populated for validation (422) failures; otherwise null/empty.
 */
public record Problem(
		String type,
		String title,
		int status,
		String code,
		String detail,
		String instance,
		String traceId,
		OffsetDateTime timestamp,
		List<FieldError> fieldErrors) {

	/** A single field-level validation error. */
	public record FieldError(String field, String code, String message) {
	}
}

File: apps/web/src/api/problem.ts
import { AxiosError } from "axios";

import type { AppError, Problem } from "../types/api";

/** Convert an unknown axios/network error into a normalized AppError. */
export function toAppError(error: unknown): AppError {
  const axiosError = error as AxiosError<Problem>;
  const data = axiosError.response?.data;

  if (data && typeof data === "object" && "code" in data) {
    return {
      status: data.status,
      code: data.code,
      message: data.detail ?? data.title,
      fieldErrors: data.fieldErrors?.map((f) => ({ field: f.field, message: f.message })),
      traceId: data.traceId,
    };
  }

  return {
    status: axiosError.response?.status ?? 0,
    code: "NETWORK_ERROR",
    message: axiosError.message ?? "Network error",
  };
}
